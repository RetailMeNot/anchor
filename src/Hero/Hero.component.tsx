// REACT
import * as React from 'react';

// VENDOR
import styled from '@xstyled/styled-components';
import classnames from 'classnames';
// ANCHOR
import {
    Typography,
    TypographyProps,
    TextAlign,
} from '../Typography/Typography.component';
import { colors } from '../theme';

// Hero
// ------------------------------------------------------------------------------------------------------------------

interface StyledHeroProps {
    align?: TextAlign;
    background?: string;
    color?: string;
    minHeight?: string;
}

const StyledHero = styled.section<StyledHeroProps>`
    background: ${props => props.background};
    box-sizing: border-box;
    position: relative;
    width: 100%;
    margin: 0 auto;
    display: block;
    color: ${({ color = colors.white.base }) => color};
    text-align: ${props => props.align};
    min-height: ${props => props.minHeight};
    display: flex;
    flex-direction: column;
    justify-content: center;
`;

interface HeroProps {
    background?: string;
    align?: TextAlign;
    color?: string;
    className?: string;
    children?: any;
    minHeight?: string;
}

export const Hero = ({
    background = 'transparent',
    align = 'center',
    minHeight = '7.5rem',
    className,
    children,
    ...props
}: HeroProps): React.ReactElement<any> => (
    <StyledHero
        className={classnames('anchor-hero', className)}
        background={background}
        align={align}
        minHeight={minHeight}
        {...props}
    >
        {children}
    </StyledHero>
);

// Hero.Title
// ------------------------------------------------------------------------------------------------------------------

const StyledTitle = styled(Typography)`
    text-shadow: 0 0.125rem 0.25rem rgba(0, 0, 0, 0.09);
    position: relative;
`;

export const HeroTitle = ({
    className,
    children,
    tag = 'h1',
    weight = 600,
    scale = 32,
    ...props
}: TypographyProps): React.ReactElement<any> => (
    <StyledTitle
        className={classnames('anchor-hero-title', className)}
        tag={tag}
        weight={weight}
        scale={scale}
        {...props}
    >
        {children}
    </StyledTitle>
);

Hero.Title = HeroTitle;

// Hero.Subtitle
// ------------------------------------------------------------------------------------------------------------------

const StyledSubtitle = styled(Typography)`
    text-shadow: 0 0.125rem 0.125rem rgba(0, 0, 0, 0.04);
    position: relative;
    display: block;
`;

export const HeroSubtitle = ({
    className,
    children,
    scale = 18,
    weight = 600,
    ...props
}: TypographyProps): React.ReactElement<any> => (
    <StyledSubtitle
        className={classnames('anchor-hero-subtitle', className)}
        scale={scale}
        weight={weight}
        {...props}
    >
        {children}
    </StyledSubtitle>
);

Hero.Subtitle = HeroSubtitle;
