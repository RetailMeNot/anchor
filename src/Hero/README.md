# Hero

The Hero component offers two built-in components `Hero.Title` and `Hero.Subtitle`, to compose along with any other desired components (e.g. containers, images, etc).

## Usage

#### Basic

```jsx
import { Hero } from '@retailmenot/anchor';
const { Title, Subtitle } = Hero;

const basicExample = props => (
    <Hero background="blue">
        <Title>Title</Title>
        <Subtitle>My Subtitle</Subtitle>
    </Hero>
);
```

#### Advanced

```jsx
import { Container, SomeImage } from 'my_app/components';
import { Hero } from '@retailmenot/anchor';
const { Title, Subtitle } = Hero;

const advancedExample = props => (
    <Hero align="left" color="neutrals.charcoal.base" minHeight="36rem">
        <Container>
            <Title>Socks with Sandals</Title>
            <Subtitle>The best thing since sliced bread!</Subtitle>
            <Button>Buy Some Today</Button>

            <SomeImage position="right" />
        </Container>
    </Hero>
);
```

## Props

### Hero

-   **`background = 'transparent'`**: css property for defining the Hero's background
-   **`align = 'center'`:** css property for text alignment
-   **`minHeight = '7.5rem'`:** css property for min-height
-   **`color = '#fff'`:** css property for text color
-   **`className`:** Additional className to be added to the Hero element

### Hero.Title / Hero.Subtitle

See `Typography` anchor component.
