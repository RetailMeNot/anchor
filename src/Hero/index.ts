export { Hero } from './Hero.component';
