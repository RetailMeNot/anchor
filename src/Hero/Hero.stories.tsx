// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
import styled from 'styled-components';
import { text, color } from '@storybook/addon-knobs';
// ANCHOR
import { Typography } from '..';
import { colors } from '../theme';
import * as README from './README.md';
import SomeImage from './temporary/some_image.svg';
import { Grid, CustomAdaptor, StandardBreakpoints } from '../Grid';
import { Button } from '../Button';

import { Hero } from './Hero.component';
const { Title, Subtitle } = Hero;

// Styled Components
// ------------------------------------------------------------------------------------------------------------------
const StyledStory = styled.div`
    background: ${colors.white.base};
    width: 100vw;
`;

// Stories
// ------------------------------------------------------------------------------------------------------------------
const tealGradient = 'linear-gradient(224deg,#00cccd,#2de2b3)';

storiesOf('Components/Hero', module)
    .addParameters({
        readme: {
            sidebar: README,
        },
    })
    .add('Default', () => (
        <StyledStory>
            <Hero
                background={color('Hero Background', tealGradient)}
                minHeight="9.375rem"
            >
                <Title scale={44}>{text('Title', 'Baratheon')}</Title>
                <CustomAdaptor minWidth={StandardBreakpoints.sm.min}>
                    <Subtitle>{text('Subtitle', 'Ours is the Fury')}</Subtitle>
                </CustomAdaptor>
            </Hero>
        </StyledStory>
    ))
    .add('Simple', () => {
        const BreadCrumbs = styled(Typography)`
            &&{ margin-bottom: 6px; }}
        `;

        return (
            <StyledStory>
                <Hero
                    background={color('Hero Background', tealGradient)}
                    color={color('Hero Color', colors.white.base)}
                >
                    <BreadCrumbs scale={14} weight={500}>
                        {text('Subtext', 'Westeros / House Tyrell')}
                    </BreadCrumbs>
                    <Title>{text('Title', 'Olenna Tyrell')}</Title>
                </Hero>
            </StyledStory>
        );
    })
    .add('CTA', () => {
        // Replace when proper button exists
        const NeatButton = styled(Button)`
            color: white;
            border: thin solid white;
            width: 9.5rem;
            text-align: center;
            margin: 0.1835rem auto;
            font-size: 15px;
        `;

        const StyledImageWrapper = styled.div`
            position: absolute;
            right: 0;
            bottom: 0;
            svg {
                display: block;
            }
        `;

        return (
            <StyledStory>
                <Hero
                    background={color('Hero Background', tealGradient)}
                    minHeight="12.5rem"
                >
                    <Grid gap="0.375rem" columns={1}>
                        <StyledImageWrapper>
                            <SomeImage />
                        </StyledImageWrapper>
                        <Title>{text('Title', 'House Lannister')}</Title>
                        <CustomAdaptor minWidth={StandardBreakpoints.sm.min}>
                            <Subtitle>
                                {text(
                                    'Subtitle',
                                    'A Lannister always pays his debts'
                                )}
                            </Subtitle>
                        </CustomAdaptor>
                        <NeatButton
                            variant="text"
                            onClick={() => alert('Paid!')}
                        >
                            {text('Button Text', 'Pay Debts Today')}
                        </NeatButton>
                    </Grid>
                </Hero>
            </StyledStory>
        );
    });
