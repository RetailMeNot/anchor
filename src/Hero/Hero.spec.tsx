// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// ENZYME
import { shallow, mount } from 'enzyme';
// COMPONENT
import {Hero} from './Hero.component';
const { Title, Subtitle } = Hero;

// TEST SETUP
const subject = (
    <Hero>
        <Title>Hero Title</Title>
        <Subtitle>Hero Subtitle</Subtitle>
    </Hero>
);
const wrapper = mount(subject);
const component = shallow(subject);

describe('Component: Hero', () => {
    it('should be defined', () => {
        /* tslint:disable */
        console.log(subject);
        console.log(wrapper.debug());
        console.log(component.debug());

        /* tslint:enable */
        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
