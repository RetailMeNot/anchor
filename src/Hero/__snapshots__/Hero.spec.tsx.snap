// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Hero should be defined 1`] = `
<section
  className="anchor-hero sc-gipzik fqnMqj"
/>
`;

exports[`Component: Hero should render with custom content 1`] = `
<section
  className="anchor-hero sc-gipzik fqnMqj"
>
  <h1
    className="anchor-typography anchor-hero-title sc-csuQGl jAXjpX sc-caSCKo jgTECj"
    scale={32}
  >
    Good Ole Title
  </h1>
  <span
    className="anchor-typography sc-gisBJw bGdbKe"
  >
    Some Typography text
  </span>
  <button
    className="anchor-button sc-gqjmRU jfRbTD"
    size="large"
  >
    CTA
  </button>
</section>
`;

exports[`Component: Hero should render with title 1`] = `
<section
  className="anchor-hero sc-gipzik fqnMqj"
>
  <h1
    className="anchor-typography anchor-hero-title sc-csuQGl jAXjpX sc-dVhcbM jMfNJo"
    scale={32}
  >
    Hero Title
  </h1>
</section>
`;

exports[`Component: Hero should render with title and subtitle 1`] = `
<section
  className="anchor-hero sc-gipzik fqnMqj"
>
  <h1
    className="anchor-typography anchor-hero-title sc-csuQGl jAXjpX sc-eqIVtm eeazzE"
    scale={32}
  >
    Hero Title
  </h1>
  <span
    className="anchor-typography anchor-hero-subtitle sc-Rmtcm KPeEa sc-fAjcbJ dBoqIT"
    scale={18}
  >
    Hero Subtitle
  </span>
</section>
`;
